File: app/app/src/main/java/williamhester/me/breadit2/ui/viewholders/SubmissionViewHolder.java
package williamhester.me.breadit2.ui.viewholders;

import android.content.res.Resources;
import android.view.View;
import android.widget.TextView;
import butterknife.ButterKnife;
import williamhester.me.breadit2.R;
import williamhester.me.breadit2.html.HtmlParser;
import williamhester.me.breadit2.models.Submission;
import williamhester.me.breadit2.ui.VotableCallbacks;

import static butterknife.ButterKnife.findById;

/**
 * Created by william on 6/13/16.
 */
public class SubmissionViewHolder extends ContentViewHolder<Submission> {

  private final TextView pointsTextView;
  private final TextView nsfwTextView;
  private final TextView titleTextView;
  private final TextView metadata1TextView;
  private final TextView metadata2TextView;

  private Submission submission;

  public SubmissionViewHolder(
      View itemView, final VotableCallbacks clickListener) {
    super(itemView);

    itemView.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        clickListener.onVotableClicked(submission);
      }
    });

    pointsTextView = ButterKnife.findById(itemView, R.id.points);
    nsfwTextView = findById(itemView, R.id.nsfw);
    titleTextView = findById(itemView, R.id.title);
    metadata1TextView = findById(itemView, R.id.metadata_1);
    metadata2TextView = findById(itemView, R.id.metadata_2);
  }

  @Override
  public void setContent(Submission item) {
    submission = item;
    Resources res = itemView.getContext().getResources();
    String pointsString = res.getQuantityString(R.plurals.points, item.getScore());
    pointsTextView.setText(String.format(pointsString, item.getScore()));

    nsfwTextView.setVisibility(item.isNsfw() ? View.VISIBLE : View.INVISIBLE);

    titleTextView.setText(item.getTitle());

    metadata1TextView.setText(String.format(res.getString(R.string.metadata_1_line),
        item.getAuthor(), item.getSubreddit(), item.getDomain()));

    String metadata2String = res.getQuantityString(R.plurals.metadata_2_line,
        item.getNumComments());
    metadata2TextView.setText(String.format(metadata2String, item.getNumComments(),
        calculateShortTime(item.getCreatedUtc())));
  }

  /** Gets the underlying content. */
  protected Submission getContent() {
    return submission;
  }
}

File: app/app/src/main/java/me/williamhester/reddit/ui/viewholders/SubmissionViewHolder.kt
package me.williamhester.reddit.ui.viewholders

import android.view.View
import android.widget.TextView
import me.williamhester.reddit.R
import me.williamhester.reddit.models.Submission
import me.williamhester.reddit.ui.VotableCallbacks

/** A ViewHolder for a Submission object */
open class SubmissionViewHolder(
    itemView: View,
    clickListener: VotableCallbacks
) : ContentViewHolder<Submission>(itemView) {

  private val pointsTextView: TextView = itemView.findViewById(R.id.points)
  private val nsfwTextView: TextView = itemView.findViewById(R.id.nsfw)
  private val titleTextView: TextView = itemView.findViewById(R.id.title)
  private val metadata1TextView: TextView = itemView.findViewById(R.id.metadata_1)
  private val metadata2TextView: TextView = itemView.findViewById(R.id.metadata_2)

  private var submission: Submission? = null

  init {
    itemView.setOnClickListener { clickListener.onVotableClicked(submission!!) }
  }

  override fun setContent(item: Submission) {
    submission = item
    val res = itemView.context.resources
    val pointsString = res.getQuantityString(R.plurals.points, item.score)
    pointsTextView.text = String.format(pointsString, item.score)

    nsfwTextView.visibility = if (item.isNsfw) View.VISIBLE else View.INVISIBLE

    titleTextView.text = item.title

    metadata1TextView.text = String.format(res.getString(R.string.metadata_1_line),
        item.author, item.subreddit, item.domain)

    val metadata2String = res.getQuantityString(R.plurals.metadata_2_line,
        item.numComments)
    metadata2TextView.text = String.format(metadata2String, item.numComments,
        calculateShortTime(item.createdUtc))
  }

  override fun getContent(): Submission? {
    return submission
  }
}
